Framework/Beamline/src/DetectorInfo.cpp

```diff
@@ -78,28 +90,250 @@ bool DetectorInfo::isEquivalent(const DetectorInfo &other) const {
+  m_positions.access()[linearIndex(index)] = position;
+}
+
+/// Set the rotation of the detector with given index.
+void DetectorInfo::setRotation(const std::pair<size_t, size_t> index,
+                               const Eigen::Quaterniond &rotation) {
+  m_rotations.access()[linearIndex(index)] = rotation.normalized();
+}
+
+/// Returns the scan count of the detector with given detector index.
+size_t DetectorInfo::scanCount(const size_t index) const {
+  if (!m_scanCounts)
+    return 1;
+  return (*m_scanCounts)[index];
+}
+
+/** Returns the scan interval of the detector with given index.
+ *
+ * The interval start and end values would typically correspond to nanoseconds
+ * since 1990, as in Kernel::DateAndTime. */
+std::pair<int64_t, int64_t>
+DetectorInfo::scanInterval(const std::pair<size_t, size_t> index) const {
+  if (!m_scanIntervals)
+    return {0, 0};
+  return (*m_scanIntervals)[linearIndex(index)];
+}
+
+/* Set the scan interval of the detector with given index.
+ *
+ * The interval start and end values would typically correspond to nanoseconds
+ * since 1990, as in Kernel::DateAndTime. */
+void DetectorInfo::setScanInterval(const std::pair<size_t, size_t> index,
+                                   std::pair<int64_t, int64_t> interval) {
+  if (!m_scanIntervals)
+    initScanIntervals();
+  // We forbid this since we (currently?) can not verify that the new interval
+  // has no collisions with any of the other intervals for this detector.
+  checkNoTimeDependence();
+  if (interval.first >= interval.second)
+    throw std::runtime_error(
+        "DetectorInfo: cannot set scan interval with start > end");
+  m_scanIntervals.access()[linearIndex(index)] = interval;
+}
+
+namespace {
+void failMerge(const std::string &what) {
+  throw std::runtime_error(std::string("Cannot merge DetectorInfo: ") + what);
+}
+
+std::pair<size_t, size_t>
+getIndex(const Kernel::cow_ptr<std::vector<std::pair<size_t, size_t>>> &indices,
+         const size_t index) {
+  if (!indices)
+    return {index, 0};
+  return (*indices)[index];
+}
+}
+
+/** Merges the contents of other into this.
+ *
+ * Scan intervals in both other and this must be set. Intervals must be
+ * identical or non-overlapping. If they are identical all other parameters (for
+ * that index) must match.
+ *
+ * Time indices in `this` are preserved. Time indices added from `other` are
+ * incremented by the scan count of that detector in `this`. The relative order
+ * of time indices added from `other` is preserved. If the interval for a time
+ * index in `other` is identical to a corresponding interval in `this`, it is
+ * ignored, i.e., no time index is added. */
+void DetectorInfo::merge(const DetectorInfo &other) {
+  if (size() != other.size())
+    failMerge("size mismatch");
+  if (!m_scanIntervals || !other.m_scanIntervals)
+    failMerge("scan intervals not defined");
+  if (!(m_isMonitor == other.m_isMonitor) &&
```

Comment:
No it is in principle the same but `cow_ptr` does not support that. I think I actually have another open PR that adds this operator, if that gets merged first we can clean this up here.